[SPARK-47208][CORE] Allow overriding base overhead memory

jpcorreia99 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
We can already select the desired overhead memory directly via the `spark.driver/executor.memoryOverhead` flags, however, if that flag is not present the overhead memory calculation goes as follows:

```
overhead_memory = Max(384, 'spark.driver/executor.memory' * 'spark.driver/executor.memoryOverheadFactor')

[where the 'memoryOverheadFactor' flag defaults to 0.1]
```

This PR adds two new spark configs: `spark.driver.minMemoryOverhead` and `spark.executor.minMemoryOverhead`, which can be used to override the 384Mib minimum value.

The memory overhead calculation will now be :

```
min_memory = sparkConf.get('spark.driver/executor.minMemoryOverhead') // has default of 384

overhead_memory = Max(min_memory, 'spark.driver/executor.memory' * 'spark.driver/executor.memoryOverheadFactor')
```

### Why are the changes needed?
There are certain times where being able to override the 384Mb minimum directly can be beneficial. We may have a scenario where a lot of off-heap operations are performed (ex: using package managers/native compression/decompression) where we don't have a need for a large JVM heap but we may still need a signficant amount of memory in the spark node. 

Using the `memoryOverheadFactor` config flag may not prove appropriate, since we may not want the overhead allocation to directly scale with JVM memory, as a cost saving/resource limitation problem.

### Does this PR introduce _any_ user-facing change?
Yes, as described above, two new flags have been added to the spark config. No break of existing behaviours.

### How was this patch tested?
Added tests for 3 cases:
- If `spark.driver/executor.memoryOverhead` is set, then the new changes have no effect.
- If  `spark.driver/executor.minMemoryOverhead` is set and its value is higher than  'spark.driver/executor.memory' * 'spark.driver/executor.memoryOverheadFactor', the total memory will be the allocated JVM memory + `spark.driver/executor.minMemoryOverhead`
- If  `spark.driver/executor.minMemoryOverhead` but its value is lower than 'spark.driver/executor.memory' * 'spark.driver/executor.memoryOverheadFactor', the total memory will be the allocated JVM memory + 'spark.driver/executor.memory' * 'spark.driver/executor.memoryOverheadFactor'.

<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->


### Was this patch authored or co-authored using generative AI tooling?
No
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->

jpcorreia99 committed:
wip

jpcorreia99 committed:
Finish implementation for yarn

jpcorreia99 committed:
add docs + fix test

jpcorreia99 committed:
Add proper spark jira link

jpcorreia99 commented on core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala:
```diff
@@ -489,10 +489,13 @@ object ResourceProfile extends Logging {
       executorMemoryMiB: Long,
       overheadFactor: Double): Long = {
+    val minMemoryOverhead =
+      minimumOverHeadMemoryFromConf.getOrElse(ResourceProfile.MEMORY_OVERHEAD_MIN_MIB);
     overHeadMemFromConf.getOrElse(math.max((overheadFactor * executorMemoryMiB).toInt,
-        ResourceProfile.MEMORY_OVERHEAD_MIN_MIB))
+      minMemoryOverhead))
```
I order to minimize the changes to existing logic, I chose to make this flag not have a default and instead write the logic to check if it is not set and use the existing constant.

However, from a code cleaniness and maintenance standpoint, I think the constant can be deleted and the flag value defaulted to 384Mib. 

@the reviewer, what's your view here?

tgravescs replied:
I think it would be cleaner just to have the defaults of the configs 384.

jpcorreia99 committed:
correct version in docs

jpcorreia99 commented on docs/configuration.md:
```diff
@@ -202,6 +202,15 @@ of the most common options to set are:
+    The minimum amount of non-heap memory to be allocated per driver process in cluster mode, in MiB unless otherwise specified, if <code>spark.driver.memoryOverhead</code> is not defined.
+    This option is currently supported on YARN and Kubernetes.
+  </td>
+  <td>3.5.2</td>
```
What value should be added here? I could not find a roadmap of future releases for spark, but saw the latest release is '3.5.1'

tgravescs replied:
oops, answered this in a different comment, should be 4.0.0

jpcorreia99 committed:
remove extra newlines

jpcorreia99 commented:
Hey @HyukjinKwon, would you happen to have the time to give this a review or point me to someone who might?

HyukjinKwon commented:
This isn't really my area. cc @tgravescs FYI

jpcorreia99 commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -117,6 +117,14 @@ package object config {
+    .doc("The minimum amount of non-heap memory to be allocated per driver in cluster mode, " +
+      "in MiB unless otherwise specified. Overrides the default value of 384Mib." +
+      " This value is ignored if spark.driver.memoryOverhead is set directly.")
+    .version("3.5.2")
```
Should this be 4.0 instead?

tgravescs replied:
this should be 4.0.0. This is a new feature so generally wouldn't be back ported, at least without discussion.

tgravescs commented:
can you provide more details as to why this is needed?  Its yet another configuration for memory overhead when we already have multiple.   I assume you aren't trying to make the minimum lower because you say using a lot of off-heap memory.  You can manually set the memoryoverhead vs setting this new config, so what use case is this trying to fix?

jpcorreia99 commented:
@tgravescs very fair point!

We preconfigure the spark nodes we provide to users with a global config, but then allow the users to override some of the fields.
We observed a reduction in OOMs when bumping the minimum memory overhead.

The reason why we don't explicitly set the memory overhead is because we could accidentally be reducing the overall memory the user has access to.

As an example, we'd like to have at least 500Mb of overhead. User has requested a 30Gb driver. If we directly set the memory overhead to 500Mb, we'd be cutting the overhead available to the user by 2.5Gb (as overhead memory is calculated as 10% of JVM).

We also don't want to be changing the overhead factor since these are global config. If we'd like to have at least a 500Mb overhead and we have a default memory of 2Gb, we'd need to set up the factor at 0.25. But then, if the user comes and asks for a 30Gb driver, he'd also receive 7.5Gb in overhead, which is excessive.

As such, defining a global overhead memory minimum would work best if you want to have configurable memory settings and reasonable overhead defaults.

tgravescs commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -358,6 +366,15 @@ package object config {
+    .doc("The minimum amount of non-heap memory to be allocated per executor " +
+      "in MiB unless otherwise specified. Overrides the default value of 384Mib." +
+      " This value is ignored if spark.executor.memoryOverhead is set directly.")
+    .version("3.5.2")
```
4.0.0

tgravescs commented on resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -75,7 +75,7 @@ private[spark] class Client(
   private val hadoopConf = new YarnConfiguration(SparkHadoopUtil.newConfiguration(sparkConf))
 
-  private val isClusterMode = sparkConf.get(SUBMIT_DEPLOY_MODE) == "cluster"
+      private val isClusterMode = sparkConf.get(SUBMIT_DEPLOY_MODE) == "cluster"
```
remove extra spaces

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala:
```diff
@@ -706,6 +706,39 @@ class ClientSuite extends SparkFunSuite
+          appContext.getResource should be(Resource.newInstance(1524L, 1))
+        } else {
+          // 512 driver default (non-cluster) + 384 overhead default
+          //  that can't be changed in non cluster mode.
```
nit extra space before "that"

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnAllocatorSuite.scala:
```diff
@@ -128,6 +128,9 @@ class YarnAllocatorSuite extends SparkFunSuite
       .set(EXECUTOR_MEMORY, 2048L)
 
+    // scalastyle:off println
+    sparkConfClone.getAll.foreach(println)
```
assume this was debugging? remove

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnAllocatorSuite.scala:
```diff
@@ -772,6 +775,40 @@ class YarnAllocatorSuite extends SparkFunSuite
+      assert(memory == (executorMemory + 100))
+    } finally {
+      sparkConf
+        .remove(EXECUTOR_MIN_MEMORY_OVERHEAD)
```
sparkConf.remove(EXECUTOR_MEMORY_OVERHEAD)

tgravescs commented on resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -96,12 +96,21 @@ private[spark] class Client(
     sparkConf.get(amMemoryOverheadEntry).getOrElse(
       math.max((amMemoryOverheadFactor * amMemory).toLong,
-        ResourceProfile.MEMORY_OVERHEAD_MIN_MIB)).toInt
+        driverMinimumMemoryOverhead)).toInt
```
driver and AM aren't necessarily the same process.  In cluster mode it is but in client mode it isn't (see spark.yarn.am.memoryOverheah) - https://spark.apache.org/docs/latest/running-on-yarn.html.
Unless you have a use case for the app master doing native stuff, which I don't see, I think this should stay the same (384) or if you have a use case for it we should have a separate config for it since its a YARN specific thing.

jpcorreia99 replied:
I may be misunderstanding something, but doesn't the fact that above we have 

```
  private val driverMinimumMemoryOverhead =
    if (isClusterMode) {
      sparkConf.get(DRIVER_MIN_MEMORY_OVERHEAD)
    } else {
      ResourceProfile.MEMORY_OVERHEAD_MIN_MIB
    }
```
    
 guarantee that my changes are only applied in cluster mode, which, form that you said, seems to be the case where it actually corresponds to the driver?

tgravescs replied:
ah yes that works

tgravescs reviewed (commented):
ok, for an overall cluster config that makes sense, but on the other hand you are increasing the minimum for everyone as well and if they don't need the extra memory its wasting it but I guess that is a trade off.  I can see the use case, left some comments

mridulm commented:
I would like to understand the usecase better here - It is still unclear to me what characteristics you are shooting for by this PR.

Reduction in OOM is mentioned [as a usecase](https://github.com/apache/spark/pull/45240#issuecomment-1976891862) - but overhead does not impact that, heap memory does.

> The reason why we don't explicitly set the memory overhead is because we could accidentally be reducing the overall memory the user has access to.

Any increase in overhead memory implies reduction in available heap memory if the max container memory has a upper limit and it is at that limit (if container memory is not limited - there is no impact of memory available to user, tune memory and overhead independently).

The reason behind the question is, I am trying to understand if the default min (which was estimated quite a while ago) needs to be relooked or not - and understanding why your deployment needs a higher min will help.

mridulm reviewed (commented):
We can simplify the PR by setting default of `spark.driver|executor.minMemoryOverhead` to 384, and replace existing use of 384 with this config.

jpcorreia99 commented:
Hey @mridulm, thanks for commenting!

>  would like to understand the usecase better here - It is still unclear to me what characteristics you are shooting for by this PR.

I recognize it's a bit of a contrived scenario. We provide users the ability to write spark code while we handle the whole infrastructure provisioning. We have defaults for both the JVM but allow the users to override it. In Spark, the default rule for off heap allocation is 10% of the JVM, with this value being configurable via the `memoryOverheadFactor` config. 

Let's now say that as part of the infrastructure provisioning we have to perform certain off-heap operations we know will need 500Mb of off-heap memory. 

We can:
- Directly set the `memoryOverhead` config. But here we risk actually giving the user less overhead memory than he would be given otherwise (e.g. if the user requested 10GB JVM, with this conf he's only get 500MB instead of the 1Gb (10*0.1) he'd have otherwise.
- We can set the `memoryOverheadFactor` to give us the desired value, but let's imagine a scenario where the default JVM allocation is 1GB: to get 500MB of off-heap we'd need a factor of 0.5. If the user then decided to request for 10GB of JVM instead, we'd be also allocating 5Gb off off-heap memory, which could be very wasteful.

As such, that's why I'm proposing this config, as a way to let off-heap memory allocation scale as it usually does but while guaranteeing that we have *at least* a certain configurable amount of off heap memory! 

> Reduction in OOM is mentioned https://github.com/apache/spark/pull/45240#issuecomment-1976891862 - but overhead does not impact that, heap memory does.

In our case the OOMs we got where K8s OOMkiller OOMs, not JVM ones. These originated from the off-heap processes consuming too much memory and k8s killing down the pod. So, in this scenario, increasing the overhead actually stopped the OOMs

> Any increase in overhead memory implies reduction in available heap memory if the max container memory has a upper limit and it is at that limit (if container memory is not limited - there is no impact of memory available to user, tune memory and overhead independently).

That's a fair concern. In our case we allocate the containers based on the memory required for the JVM + overhead instead of  having a fixed memory container, so we don't run into this issue. 

> The reason behind the question is, I am trying to understand if the default min (which was estimated quite a while ago) needs to be relooked or not - and understanding why your deployment needs a higher min will help.

The 387mb value does seem a bit arbitrary 😅 but it seems mostly fine. In our case we simply rely on heavy off-heap operations. Some examples would be package managers like conda + pip, performing  ZSTD decompression (the jni implementation delegates to native operations), even up until spark 3.2, the `SparkContext::AddArchive` call would launch `tar` processes for decompression.

jpcorreia99 committed:
address comments

jpcorreia99 commented:
Hey @tgravescs I've now addressed your comments and also answered @mridulm, whenever you have the time, I'd highly appreciate if you could give it another review.

jpcorreia99 commented on docs/configuration.md:
```diff
@@ -202,6 +202,15 @@ of the most common options to set are:
 </tr>
+<tr>
+  <td><code>spark.driver.minMemoryOverhead</code></td>
+  <td>384Mib</td>
```
@tgravescs just want to make sure I'm being correct here.
I create this flag with "  .createWithDefaultString("384m")". Is it correct for me to write here `Mib` instead of `m` for clarity to the reader or should I stick to exactly the value it contains?

mridulm reviewed (commented):
Looks good to me, just one point for discussion

jpcorreia99 commented on core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala:
```diff
@@ -489,10 +489,11 @@ object ResourceProfile extends Logging {
       overheadFactor: Double): Long = {
     overHeadMemFromConf.getOrElse(math.max((overheadFactor * executorMemoryMiB).toInt,
-        ResourceProfile.MEMORY_OVERHEAD_MIN_MIB))
+      minimumOverHeadMemoryFromConf))
```
I do agree, at this point, outside of tests, `ResourceProfile.MEMORY_OVERHEAD_MIN_MIB` has only one usage in client AM overhead, so will make that change, can revert if @tgravescs disagrees

jpcorreia99 replied:
Addressed in 3abd57efffcc3df2e2f4225b45f00f5e8af01724

tgravescs replied:
I agree with removing it.

jpcorreia99 committed:
Remove ResourceProfile.MEMORY_OVERHEAD_MIN_MIB

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala:
```diff
@@ -706,6 +706,39 @@ class ClientSuite extends SparkFunSuite
+
+        appContext.getApplicationName should be("foo-test-app")
+        // flag should only work for cluster mode
+        if (deployMode=="cluster") {
```
nit spaces around " == "

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala:
```diff
@@ -706,6 +706,39 @@ class ClientSuite extends SparkFunSuite
+        // flag should only work for cluster mode
+        if (deployMode=="cluster") {
+          // 1Gb driver default + 500 overridden minimum default overhead
+          appContext.getResource should be(Resource.newInstance(1524L, 1))
```
nit space after "should be "  for readability

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnAllocatorSuite.scala:
```diff
@@ -772,6 +773,42 @@ class YarnAllocatorSuite extends SparkFunSuite
+      val memory = defaultResource.getMemorySize
+      assert(memory == (executorMemory + 100))
+    } finally {
+      sparkConf
```
nit the .remove should be on same line as sparkConf as its not to long

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnAllocatorSuite.scala:
```diff
@@ -772,6 +773,42 @@ class YarnAllocatorSuite extends SparkFunSuite
+      sparkConf
+        .remove(EXECUTOR_MIN_MEMORY_OVERHEAD)
+      sparkConf
+        .remove(EXECUTOR_MIN_MEMORY_OVERHEAD)
```
you are removing EXECUTOR_MIN_MEMORY_OVERHEAD twice, one of these should be EXECUTOR_MEMORY_OVERHEAD right?

tgravescs commented on resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnAllocatorSuite.scala:
```diff
@@ -772,6 +773,42 @@ class YarnAllocatorSuite extends SparkFunSuite
+      val memory = defaultResource.getMemorySize
+      assert(memory == (executorMemory + 500))
+    } finally {
+      sparkConf
```
nit put the .removeConf on the same line as sparkConf

jpcorreia99 committed:
address formatting

jpcorreia99 commented:
@tgravescs great! Any timeline on merging this?

tgravescs commented:
Merged to master, thanks @mridulm  @jpcorreia99
